doi/retro_doi_builder.py

```diff
@@ -0,0 +1,467 @@
+MILLIS_PER_DAY = 24 * 60 * 60 * 1000
+
+DATASET_L0_PROVENANCE_COLUMNS = [
+        'subsite', 'node', 'sensor', 'method', 'deployment', 'id', 'fileName',
+        'parserName', 'parserVersion']
+ALL_DATASET_L0_PROVENANCE = 'SELECT {0} FROM dataset_l0_provenance'.format(
+        ','.join(DATASET_L0_PROVENANCE_COLUMNS))
+
+STREAM_METADATA_SERVICE_URL_TEMPLATE = 'http://{0}:12571/streamMetadata'
+PARTITION_METADATA_SERVICE_URL_TEMPLATE = 'http://{0}:12571/partitionMetadata'
+XASSET_SERVICE_URL_TEMPLATE = 'http://{0}:12587/'
+DOI_SERVICE_URL_TEMPLATE = 'http://{0}:12588/doi'
+PARSED_URL_TEMPLATE = """%s/sensor/inv/%s/%s/%s/%s/%s?beginDT=%s&endDT=%s&
+                      execDPA=false&user=doilookup"""
+DATE_REGEX = re.compile("""\\d{4}[_-]?(0[1-9]|1[0-2])[-_]?(0[1-9]|[12][0-9]|
+                        3[01])""")
+
+
+# Take a timestamp representing milliseconds since the UNIX epoch and return
+# an ISO8601 string representation of that time. For example, an integer input
+# of 1494524856000 would result in a string output of "2017-05-11T17:47:36".
+def convertMillisToIso8601(timestamp):
+        return datetime.utcfromtimestamp(timestamp/1000).isoformat()
+
+
+def streamed(method):
+    return 'streamed' in method
+
+
+def createDOIRecordsHelper(params):
+    (builder, row) = params
+    createDOIRecords(builder, row)
+
+
+def createDOIRecords(builder, row):
+    (subsite, node, sensor, method, deploymentNumber, parsername,
+     parserversion, version, start, stop, latitude, longitude, depth,
+     manufacturer, model, serialnumber, mio, name, uid, rawStreams) = row
+    if mio is None or mio == "":
+        mio = "OOI - Ocean Observatories Initiative"
+    streams = json.loads(rawStreams)
+    refdes = "-".join((subsite, node, sensor))
+
+    # make parser DOI
+    parserDOI = {}
+    parserDOI["@class"] = ".ParserDOIRecord"
+    parserDOI["doi"] = None
+    parserDOI["obsolete"] = False
+    parserDOI["error"] = None
+    parserDOI["doiType"] = "PARSER"
+    parserDOI["name"] = parsername
+    parserDOI["version"] = parserversion
+    parserDOI["language"] = "Python"
+    parserDOI["predecessor"] = None
+    # try to create parserDOI and store the JSON response
+    try:
+        response = builder.doi_service_api.create_doi_record(parserDOI)
+        parserDOI["id"] = response["id"]
+    except DOIServiceException as e:
+        pass
+    # if the creation succeeded or not, log the result
+    builder.log_result(json.dumps(parserDOI), response["statusCode"])
+
+    # make raw DOI
+    rawFileSetDOI = {}
+    rawFileSetDOI["@class"] = ".RawFileSetDOIRecord"
+    rawFileSetDOI["doi"] = None
+    rawFileSetDOI["obsolete"] = False
+    rawFileSetDOI["error"] = None
+    rawFileSetDOI["doiType"] = "RAW_FILE_SET"
+    rawFileSetDOI["sensorDOI"] = None
+    rawFileSetDOI["referenceDesignator"] = refdes
+    rawFileSetDOI["deployment"] = deploymentNumber
+    rawFileSetDOI["method"] = method
+
+    # TODO update URL structure once it is finalized
```

Comment:
I wish I knew. The issue is that the system is depending on "splash pages" to point to lists of raw files on the Rutgers file server. The URL structures currently provided are the expected structures, but until the splash pages are finalized they could be subject to change. I need to follow up with Bill Kish on this.